File: expo-workouts/ios/GjKitWorkoutsModule.swift
// @gj-kit/expo-workouts — the Expo module (mission section 4.1, design sections 3.1 to 3.3).
//
// Every member of the native seam declared in `src/core/native-contract.ts` exists here with its
// final signature, and every one of them is a thin adapter: decode the flat bridge dictionary, call
// one `HealthStoring` member, encode the flat result, and turn any error into one of the 14 typed
// exceptions. There is no HealthKit call in this file at all — that is what keeps the whole module
// drivable from XCTest against an in-memory conformer (index f56).
//
// ── Two rules that are not optional even in a skeleton ──────────────────────────────────────────
//
// 1. EVERY member below uses the Swift-concurrency `AsyncFunction` overload — the `async throws`
//    closure, resolved by the compiler to `ConcurrentFunctionDefinition`. The closure-style overload
//    runs its body on one process-wide serial queue (index f8), where a 36 000-point route read
//    (1.1 s, f79) would block every other module in the app. HealthKit work must never go there, so
//    no member here takes a `Promise` parameter.
//
// 2. The module never talks to `HKHealthStore` directly. It holds a `HealthStoring` obtained from
//    `WorkoutsStoreInjection`, so XCTest drives this exact code against an in-memory conformer with
//    no device and no entitlement (index f56).
//
// Naming is load-bearing and is asserted by `naming-guard`: the native module string is
// `GjKitWorkouts`, the Swift class is `GjKitWorkoutsModule`, and the Kotlin class is
// `kit.gj.workouts.GjKitWorkoutsModule`.

import ExpoModulesCore
import HealthKit
import UIKit

public final class GjKitWorkoutsModule: Module {
  public func definition() -> ModuleDefinition {
    // Resolved once per module instance. No `self` is captured by any closure below: the factory
    // result is a `Sendable` existential and the `AsyncFunction` overload takes a `@Sendable`
    // closure, so capturing the module itself would be wrong as well as unnecessary.
    let store = WorkoutsStoreInjection.shared.make()

    Name("GjKitWorkouts")

    // ── availability and authorization ─────────────────────────────────────────────────────────

    // REAL. This is the one answer that must be true in the skeleton, because it is what tells a
    // consuming app whether to show the feature at all. `false` on iPad and anywhere else without a
    // health store (index f19, design section 5.7 row 1). iOS never reports `platformTooOld`
    // (the deployment target rules it out) and never `updateRequired` (no provider to update).
    AsyncFunction("availability") { () async throws -> [String: Any] in
      let availability = store.isHealthDataAvailable()
        ? AvailabilityDTO.available
        : AvailabilityDTO.notSupported
      return availability.toDictionary()
    }

    // REAL, and deliberately judgement-free: it reports raw facts and `./core` derives every
    // `ScopeStatus` from them.
    //
    // `granted` holds SHARE-authorized type identifiers only. HealthKit never discloses read
    // authorization — that is the whole reason every iOS read scope is reported as `'unknown'`
    // forever and a denied read yields an empty result instead of an error (index f14).
    AsyncFunction("authorizationSnapshot") { () async throws -> [String: Any] in
      guard store.isHealthDataAvailable() else {
        return AuthorizationSnapshotDTO(
          granted: [],
          declared: [],
          wouldPrompt: false,
          availability: .notSupported
        ).toDictionary()
      }

      let allIdentifiers = WorkoutsScope.allCases.flatMap { WorkoutsScopeTypes.identifiers[$0] ?? [] }
      let statuses = store.sharingStatus(for: allIdentifiers)
      let granted = statuses
        .filter { $0.value == HKAuthorizationStatus.sharingAuthorized.rawValue }
        .map { $0.key }
        .sorted()

      // `wouldPrompt` comes from `getRequestStatusForAuthorization` and NOTHING else — inferring it
      // from `authorizationStatus` is exactly the mistake index f14 documents.
      //
      // It asks about the READ direction ONLY, and that is load-bearing rather than a shortcut. The
      // share direction already has per-type truth in `statuses` below, while READ has none at all
      // (index f14) — `wouldPrompt` is the single bit `./core` turns into `'undetermined'` ("ask")
      // versus `'unknown'` ("the platform will never say"). Including the share side made that bit
      // STICKY TRUE, measured on device: an app that asks to write four scopes but read seven never
      // asks to SHARE heart rate or steps, so a sheet would always still be pending for those two,
      // and every read scope stayed `'undetermined'` forever even after the user granted everything.
      let identifierSet = Set(allIdentifiers)
      let wouldPrompt = (try? await store.wouldPrompt(read: identifierSet, share: [])) ?? true

      // On iOS there is no per-type manifest declaration to enumerate: authorization is gated by
      // the two Info.plist usage-description keys plus the HealthKit entitlement, and a missing key
      // CRASHES inside `requestAuthorization` rather than failing (index f19). The config plugin is
      // the structural defence and the plugin introspect snapshot asserts both keys exist, so the
      // honest answer here is "every type this build knows about".
      // `authorizationStatus(for:)` per identifier, so `./core` can tell a SHARE denial apart from
      // "never asked". `HKAuthorizationStatus`: 0 notDetermined, 1 sharingDenied, 2 sharingAuthorized.
      // Without this, `write.*` could only ever be `'granted'` or a permanent `'undetermined'`, and
      // a settings screen could never honestly offer `openSettings()`.
      var reduced: [String: String] = [:]
      for (identifier, raw) in statuses {
        reduced[identifier] =
          raw == HKAuthorizationStatus.sharingAuthorized.rawValue
            ? "granted"
            : raw == HKAuthorizationStatus.sharingDenied.rawValue ? "denied" : "undetermined"
      }

      return AuthorizationSnapshotDTO(
        granted: granted,
        declared: allIdentifiers.sorted(),
        wouldPrompt: wouldPrompt,
        statuses: reduced,
        availability: .available
      ).toDictionary()
    }

    // The seam takes TWO sets since Phase 3 (defect B). That is not a nicety: the SAME HealthKit
    // type identifier serves read and share, so one flat array cannot express what
    // `requestAuthorization(toShare:read:)` needs — it could only over-request share access or
    // silently drop it. `./core` (`iosRequestIdentifiers`) builds both sets from the public
    // `AuthorizationRequest`, so the public surface is unchanged.
    //
    // `conclusive` is always `true` on iOS: the empty-result case that forces `conclusive: false`
    // is the Android permission contract bouncing off onboarding (f120), which has no iOS twin.
    AsyncFunction("requestPermissions") { (request: [String: [String]]) async throws -> [String: Any] in
      let read = Set(request["read"] ?? [])
      let share = Set(request["write"] ?? [])
      let watched = Array(read.union(share)).sorted()

      func shareAuthorized() -> [String] {
        store.sharingStatus(for: watched)
          .filter { $0.value == HKAuthorizationStatus.sharingAuthorized.rawValue }
          .map { $0.key }
          .sorted()
      }

      let before = shareAuthorized()
      do {
        try await store.requestAuthorization(read: read, share: share)
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "requestPermissions")
      }
      return PermissionOutcomeDTO(
        before: before,
        after: shareAuthorized(),
        conclusive: true
      ).toDictionary()
    }

    // REAL. FNV-1a over the sorted granted permission strings, byte-identical to
    // `scopeFingerprint` in `src/core/sync/cursor.ts` — the two must agree or every cursor
    // resets on every call.
    AsyncFunction("grantedScopeFingerprint") { () async throws -> String in
      guard store.isHealthDataAvailable() else {
        return WorkoutsScopeFingerprint.compute([])
      }
      let allIdentifiers = WorkoutsScope.allCases.flatMap { WorkoutsScopeTypes.identifiers[$0] ?? [] }
      let granted = store.sharingStatus(for: allIdentifiers)
        .filter { $0.value == HKAuthorizationStatus.sharingAuthorized.rawValue }
        .map { $0.key }
      return WorkoutsScopeFingerprint.compute(granted)
    }

    // ── read primitives ────────────────────────────────────────────────────────────────────────

    AsyncFunction("readWorkoutPage") { (query: [String: Any]) async throws -> [String: Any] in
      // Design section 8.7: a `.strictStartDate` window query, the three-tier totals ladder per
      // workout, and the `predicateForObjects(from:)` provenance discriminator all live behind
      // `readWorkoutWindow` (RESULTS 205 to 206, f71, f73, f74, f75, f87).
      guard let window = WindowQueryDTO(dictionary: query) else {
        throw WorkoutsInvalidArgumentException("Invalid value for field: window")
      }
      let pageSize = (query["pageSize"] as? NSNumber)?.intValue ?? 0
      let pageToken = query["pageToken"] as? String
      do {
        return try await store.readWorkoutWindow(window, limit: pageSize, cursor: pageToken).toDictionary()
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "readWorkoutPage")
      }
    }

    AsyncFunction("readMetricRecords") { (query: [String: Any]) async throws -> [[String: Any]] in
      // Design section 8.4: ONE query per metric type per PAGE WINDOW, tagged with the source
      // bundle identifier so `./core` sums per origin instead of double-counting two apps that both
      // recorded the same run.
      guard let window = WindowQueryDTO(dictionary: query),
            let rawType = query["type"] as? String,
            let type = MetricTypeDTO(rawValue: rawType)
      else {
        throw WorkoutsInvalidArgumentException("Invalid value for field: metricQuery")
      }
      let origins = Set(query["origins"] as? [String] ?? [])
      do {
        return try await store.readMetricRecords(window, type: type, origins: origins).map { $0.toDictionary() }
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "readMetricRecords")
      }
    }

    AsyncFunction("readHeartRateSamples") { (query: [String: Any]) async throws -> [[String: Any]] in
      // A `.strictStartDate` window query over heart rate. Hygiene (window bounds and the 1..300 bpm
      // range) belongs to `./core` so both platforms drop the same samples.
      guard let window = WindowQueryDTO(dictionary: query) else {
        throw WorkoutsInvalidArgumentException("Invalid value for field: window")
      }
      do {
        return try await store.readHeartRateSamples(window).map { $0.toDictionary() }
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "readHeartRateSamples")
      }
    }

    AsyncFunction("hasAssociatedSamples") { (nativeId: String, quantity: String) async throws -> Bool in
      // RESULTS 206 / f71. The ONLY discriminator between a tier-1 value that is genuinely
      // `associated` and one HealthKit synthesised from a legacy workout's deprecated totals.
      guard let uuid = UUID(uuidString: nativeId), let kind = QuantityKind(rawValue: quantity) else {
        throw WorkoutsInvalidArgumentException("Invalid value for field: nativeId")
      }
      do {
        return try await store.hasAssociatedSamples(workoutUUID: uuid, quantity: kind)
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "hasAssociatedSamples")
      }
    }

    // ── sync primitives ────────────────────────────────────────────────────────────────────────

    AsyncFunction("takeCheckpoint") { () async throws -> String in
      // Design section 4.4: an `HKAnchoredObjectQueryDescriptor` run for its ANCHOR only, with no
      // results consumed (`limit: 0` means exactly that to the adapter). That is what makes the
      // initial backfill provably gap-free — the checkpoint is taken BEFORE the first page is read,
      // so anything written during the backfill still shows up in the first drain.
      do {
        return try await store.drainWorkouts(anchor: nil, limit: 0).anchor
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "takeCheckpoint")
      }
    }

    AsyncFunction("drainCheckpoint") { (checkpoint: String, limit: Int) async throws -> [String: Any] in
      // Unarchive the anchor (index f17), run the anchored query, and return the added and deleted
      // objects plus the NEW anchor. iOS anchors never expire, so `expired` is always false here —
      // only Health Connect's change token expires (index f38).
      do {
        return try await store.drainWorkouts(anchor: checkpoint, limit: limit).toDictionary()
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "drainCheckpoint")
      }
    }

    // ── routes ─────────────────────────────────────────────────────────────────────────────────

    AsyncFunction("openRoute") { (nativeId: String, consent: String) async throws -> [String: Any] in
      // Parks a `HKWorkoutRouteQueryDescriptor(route).results(for:)` iterator in a handle table.
      // `consent` is accepted for signature parity and ignored: iOS has no per-route consent step,
      // so `state` is `available` when a route sample exists and `none` when none does (index f13).
      // `consentRequired` is an Android-only state (f114, f118).
      _ = consent
      guard let uuid = UUID(uuidString: nativeId) else {
        throw WorkoutsInvalidArgumentException("Invalid value for field: nativeId")
      }
      do {
        return try await store.openRoute(workoutUUID: uuid).toDictionary()
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "openRoute")
      }
    }

    AsyncFunction("readRouteChunk") { (handle: String, maxPoints: Int) async throws -> [[String: Any]]? in
      // Converts and RELEASES each chunk; the whole `[CLLocation]` array is never accumulated
      // (415 B/point against 16 B/point converted — a 26x peak, f78). `nil` ends the stream.
      do {
        let chunk = try await store.readRouteChunk(RouteHandle(id: handle, state: .available), maxPoints: maxPoints)
        return chunk?.map { $0.toDictionary() }
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "readRouteChunk")
      }
    }

    AsyncFunction("closeRoute") { (handle: String) async throws -> Void in
      // A `break` out of the JavaScript `for await` lands here, so this must be safe to call twice
      // and safe to call with a handle that was never opened. It never throws.
      await store.closeRoute(RouteHandle(id: handle, state: .available))
    }

    // ── write primitives ───────────────────────────────────────────────────────────────────────

    AsyncFunction("findBySyncIdentifier") { (clientId: String) async throws -> [String: Any]? in
      // Design section 8.1 step 2. This lookup is what makes a re-save idempotent — re-saving at an
      // EQUAL version mints a new uuid and orphans the previous samples and route (index f26), so a
      // crash retry would otherwise become data corruption.
      do {
        return try await store.findWorkout(syncIdentifier: clientId)?.toDictionary()
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "findBySyncIdentifier")
      }
    }

    AsyncFunction("saveWorkout") { (spec: [String: Any]) async throws -> [String: Any] in
      // Design section 8.1 path B in full — pre-validate, protected-data pre-check, workout builder,
      // then a SEPARATE fresh route builder. The series builder is never used:
      // one insert error there yields a route-less workout with no error at finish (f64),
      // discarding it destroys the workout save itself (f65), and it cannot attach to an
      // already-saved workout, which the `pendingUnlock` retry requires (f66).
      guard let write = WorkoutWriteDTO(dictionary: spec) else {
        throw WorkoutsInvalidArgumentException("Invalid value for field: workoutWrite")
      }
      do {
        return try await store.saveWorkout(write).toDictionary()
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "saveWorkout")
      }
    }

    AsyncFunction("readBackVersion") { (clientId: String) async throws -> Int? in
      // Android-only by contract (f93, f94): Health Connect silently no-ops a lower client record
      // version, so that platform must always read back. HealthKit reports a stale version instead
      // of swallowing it, and our own pre-lookup catches it first (index f26). `nil` is the correct,
      // permanent iOS answer — not a stub.
      _ = clientId
      return nil
    }

    AsyncFunction("deleteWorkout") { (ref: [String: Any]) async throws -> Bool in
      // Design section 8.6: associated samples and the route sample go FIRST, then the workout, so a
      // partial failure can never leave an orphan behind. `{ clientId }` resolves through the sync
      // identifier — that one step is what stops a clientId delete from leaving metric orphans.
      let reference = DeleteRefDTO(dictionary: ref)
      do {
        if let nativeId = reference.nativeId {
          guard let uuid = UUID(uuidString: nativeId) else {
            throw WorkoutsInvalidArgumentException("Invalid value for field: nativeId")
          }
          return try await store.deleteWorkoutAndAssociated(uuid: uuid)
        }
        guard let clientId = reference.clientId, !clientId.isEmpty else {
          throw WorkoutsInvalidArgumentException("Invalid value for field: workoutRef")
        }
        guard let existing = try await store.findWorkout(syncIdentifier: clientId),
              let uuid = UUID(uuidString: existing.nativeId)
        else {
          // An unknown id is `{ deleted: false }` on both platforms, never a throw.
          return false
        }
        return try await store.deleteWorkoutAndAssociated(uuid: uuid)
      } catch {
        throw WorkoutsExceptionMapper.exception(for: error, member: "deleteWorkout")
      }
    }

    // ── platform integration ───────────────────────────────────────────────────────────────────

    AsyncFunction("openSettings") { () async throws -> Void in
      // The app's own Settings page, opened on the main actor and guarded by `canOpenURL`.
      //
      // ⚠ Asymmetry the README carries: on iOS the per-type health toggles do NOT live here. They
      // live in the Health app under Sources, and no URL scheme opens that page — so this is the
      // closest honest destination, not an equivalent of Android's Health Connect settings screen.
      try await MainActor.run {
        guard let url = URL(string: UIApplication.openSettingsURLString),
              UIApplication.shared.canOpenURL(url)
        else {
          throw WorkoutsUnavailableException("This runtime cannot open the app settings screen.")
        }
        UIApplication.shared.open(url)
      }
    }

    AsyncFunction("openStoreListing") { () async throws -> Void in
      // Android-only by contract: it exists to send the user to the Play listing for a Health
      // Connect provider update, and iOS has no provider to update. Raising `unavailable` is the
      // permanent iOS answer — not a stub.
      throw WorkoutsUnavailableException("There is no health data provider to update on this platform.")
    }
  }
}

File: expo-workouts/android/src/main/java/kit/gj/workouts/GjKitWorkoutsModule.kt
package kit.gj.workouts

import android.content.Context
import android.os.Build
import expo.modules.kotlin.activityresult.AppContextActivityResultLauncher
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * `@gj-kit/expo-workouts`의 Android 네이티브 모듈.
 *
 * 네이티브 모듈 문자열은 **`GjKitWorkouts`**이고 JS 쪽 유일한 진입점은
 * `requireOptionalNativeModule('GjKitWorkouts')`다(설계 §3.1). 클래스 이름 · 패키지 · namespace는
 * 미션 §4.1의 `naming-guard`가 문자열로 단언한다.
 *
 * ── 계약 ──────────────────────────────────────────────────────────────────────
 * 19개 seam 함수가 전부 실물이다. 이 클래스가 하는 일은 **Record -> DTO 변환과 위임**뿐이며, 규칙은
 * 두 층 아래에 산다: 순수 오케스트레이션은 `WorkoutsOperations`(JUnit이 페이크 게이트웨이로 돌린다),
 * androidx를 만지는 것은 `RealHealthConnectGateway` 하나다. 판정(활동 매핑 · 루트 위생 · 커서 ·
 * 합산 · 창 검증)은 전부 `./core`의 순수 TS다 — 네이티브는 **원시 사실만 보낸다**.
 *
 * ── 모든 AsyncFunction이 `Coroutine`인 이유 ──────────────────────────────────
 * 클로저 스타일 `AsyncFunction`은 프로세스 전역 직렬 큐에서 돈다(index f8). 36 000포인트 루트
 * 읽기(1.1 s, f79)나 41.6 s짜리 권한 온보딩(f120)이 거기 들어가면 앱 전체가 멈춘다.
 */
class GjKitWorkoutsModule : Module() {

  /** `RegisterActivityContracts`가 채운다. 등록 전에는 null이며 그 상태는 `cancelled`로 정착한다. */
  private var permissionLauncher: AppContextActivityResultLauncher<HashSet<String>, Set<String>>? = null
  private var routeConsentLauncher: AppContextActivityResultLauncher<String, RouteOutcomeDto>? = null

  private val gateway: HealthConnectGateway by lazy {
    RealHealthConnectGateway(
      context = requireContext(),
      permissionLauncher = { permissionLauncher },
      routeConsentLauncher = { routeConsentLauncher },
    )
  }

  /**
   * seam 위의 순수 오케스트레이션(§8.4·§8.5·§8.6). 이 모듈은 Record -> DTO 변환과 위임만 한다 —
   * 규칙이 여기 있으면 JUnit이 그것을 볼 수 없다.
   */
  private val operations: WorkoutsOperations by lazy {
    WorkoutsOperations(
      gateway = gateway,
      ownPackageName = requireContext().packageName,
      foregroundImportance = RealHealthConnectGateway.FOREGROUND_IMPORTANCE,
    )
  }

  private fun requireContext(): Context =
    appContext.reactContext ?: throw WorkoutsUnavailableException("noReactContext")

  override fun definition() = ModuleDefinition {
    Name("GjKitWorkouts")

    /**
     * index f9 — Android 액티비티 결과는 여기서 등록하고 `suspend launch(input)`으로 기다린다.
     * 두 컨트랙트 다 Phase 2에서 **실제로 등록된다**. 배선이 컴파일되고 등록까지 도달하는 것이
     * 이 레인의 위험한 부분이었고, 그것을 Phase 3로 미루지 않는다.
     */
    RegisterActivityContracts {
      permissionLauncher = registerForActivityResult(HealthPermissionsContract())
      routeConsentLauncher = registerForActivityResult(ExerciseRouteConsentContract())
    }

    // ── 가용성 · 인가 (데이터 읽기 예산을 소비하지 않는다) ────────────────────

    /**
     * **실제 구현이다.** `getSdkStatus(context)` tri-state + API < 28 단락 + Play 딥링크.
     * `PackageManager`는 이 경로에 오지 않는다 — f88이 Health Connect가 완벽히 동작하는 기기에서
     * PackageManager가 "provider 없음"을 보고하는 것을 측정했다.
     */
    AsyncFunction("availability") Coroutine { ->
      WorkoutsAvailability.classify(Build.VERSION.SDK_INT, gateway.sdkStatus()).toMap()
    }

    /**
     * 원시 사실만 넘긴다. `Record<Scope, ScopeStatus>` 판정은 전부 `./core`가 한다(설계 §8.8).
     * `wouldPrompt`는 iOS 전용이라 언제나 false다.
     */
    AsyncFunction("authorizationSnapshot") Coroutine { ->
      val availability = WorkoutsAvailability.classify(Build.VERSION.SDK_INT, gateway.sdkStatus())
      if (availability !is AvailabilityDto.Available) {
        // 스토어를 열 수 없으면 부여 집합을 물어볼 수도 없다. 빈 사실이 정직한 답이다.
        AuthorizationSnapshotDto(
          availability = availability,
          granted = emptyList(),
          declared = gateway.declaredHealthPermissions().sorted(),
          foreground = isForeground(),
          routeAccess = "perRoute",
          history = false,
        ).toMap()
      } else {
        val granted = gateway.grantedPermissions()
        val foreground = isForeground()
        AuthorizationSnapshotDto(
          availability = availability,
          granted = granted.sorted(),
          declared = gateway.declaredHealthPermissions().sorted(),
          foreground = foreground,
          routeAccess = routeAccess(granted, foreground),
          history = granted.contains(WorkoutsPermissions.READ_HEALTH_DATA_HISTORY),
        ).toMap()
      }
    }

    /**
     * **내부 타임아웃 없음**(f120, f122). 원시 before/after 집합을 그대로 돌려준다.
     *
     * ⚠ Phase 3 결함 B로 시그니처가 `{ read, write }`로 바뀌었다 — iOS가 read/share를 구분해야
     *   하기 때문이다(§3.2). Android에서는 방향이 권한 문자열에 이미 있으므로 합집합이 요청이고,
     *   `requested()`가 그 합집합을 준다.
     */
    AsyncFunction("requestPermissions") Coroutine { request: PermissionRequestRecord ->
      gateway.requestPermissions(request.requested()).toMap()
    }

    /**
     * 커서의 `g` 지문 재료(§4.2). 부여된 **권한 문자열**의 정렬 목록이지 scope 이름이 아니다 —
     * 그래서 어휘가 바뀌어도 기존 커서가 무효화되지 않는다(§8.8 마지막 문단).
     * FNV-1a 자체는 TS가 계산한다.
     */
    AsyncFunction("grantedScopeFingerprint") Coroutine { ->
      WorkoutsPermissions.grantedFingerprintSource(gateway.grantedPermissions())
    }

    // ── 읽기 원시 연산 ────────────────────────────────────────────────────────

    AsyncFunction("readWorkoutPage") Coroutine { query: WorkoutPageQueryRecord ->
      operations.readWorkoutPage(query.window(), query.pageSize, query.pageToken).toMap()
    }

    AsyncFunction("readMetricRecords") Coroutine { query: MetricQueryRecord ->
      gateway
        .readMetricRecords(query.metricType(), query.window(), query.origins.toSet())
        .map { it.toMap() }
    }

    AsyncFunction("readHeartRateSamples") Coroutine { window: WindowRecord ->
      gateway.readHeartRateRecords(window.toDto()).map { it.toMap() }
    }

    /**
     * **iOS 전용 판별자다**(RESULTS 206 / f71 — tier 1의 값이 합성된 legacy total인지 구별한다).
     * Health Connect에는 "워크아웃에 연관된 샘플"이라는 개념이 없다. 세션과 메트릭 레코드는 시간과
     * dataOrigin으로만 이어진다. `false`가 정직한 답이며 스텁이 아니다 — `./core`는 Android에서
     * provenance를 이 값으로 판정하지 않는다.
     */
    AsyncFunction("hasAssociatedSamples") Coroutine { _: String, _: String ->
      false
    }

    // ── 동기화 원시 연산 ──────────────────────────────────────────────────────

    AsyncFunction("takeCheckpoint") Coroutine { ->
      gateway.changesToken()
    }

    AsyncFunction("drainCheckpoint") Coroutine { checkpoint: String, limit: Int ->
      operations.drainCheckpoint(checkpoint, limit).toMap()
    }

    // ── 루트: pull 기반 핸들 ──────────────────────────────────────────────────

    /**
     * 설계 §8.4 · f113 · f116 · f118. 규칙은 전부 `WorkoutsOperations.openRoute`에 있다 —
     * 페이지 캐시 우선 · 인라인 결과 · 전경 전제 · `'prompt'`일 때만 컨트랙트.
     */
    AsyncFunction("openRoute") Coroutine { nativeId: String, consent: String ->
      operations.openRoute(nativeId, consent).toMap()
    }

    /**
     * 핸들 뒤의 스냅샷에서 `maxPoints`씩 꺼낸다. 더 없으면 **null**(스트림 종료)이다.
     * TS의 `for await`에서 나온 `break`가 `closeRoute`로 매핑된다.
     */
    AsyncFunction("readRouteChunk") Coroutine { handle: String, maxPoints: Int ->
      operations.readRouteChunk(handle, maxPoints)?.map { it.toMap() }
    }

    /** 핸들의 스냅샷을 놓는다. 이미 닫힌 핸들에 대해 조용하다(멱등) — 던지면 원래 에러가 가려진다. */
    AsyncFunction("closeRoute") Coroutine { handle: String ->
      operations.closeRoute(handle)
    }

    // ── 쓰기 원시 연산 ────────────────────────────────────────────────────────

    /**
     * iOS 전용이다(index f26 — 같은 version 재저장이 새 uuid를 만들어 이전 샘플·루트를 고아로
     * 만든다). Health Connect는 `clientRecordId` 업서트가 그 문제를 구조적으로 갖지 않으므로
     * **언제나 null**이다. 설계 §3.2가 명시한 값이며 스텁이 아니다.
     */
    AsyncFunction("findBySyncIdentifier") Coroutine { _: String ->
      NO_EXISTING_WORKOUT
    }

    AsyncFunction("saveWorkout") Coroutine { spec: WorkoutWriteRecord ->
      operations.saveWorkout(spec.toDto()).toMap()
    }

    /** **항상** 불린다. `verifyWrite` 노브는 존재하지 않는다(소유자 결정 ④, f93/f94). */
    AsyncFunction("readBackVersion") Coroutine { clientId: String ->
      operations.readBackVersion(clientId)?.toDouble()
    }

    /**
     * §8.6. `{ nativeId }` 경로는 먼저 세션을 조회해 소유권과 clientRecordId를 확인해야 메트릭
     * 고아가 생기지 않는다. UUID 형식 검증은 `./core`가 이미 끝냈다(f96).
     */
    AsyncFunction("deleteWorkout") Coroutine { ref: DeleteRefRecord ->
      operations.deleteWorkout(DeleteRefDto(nativeId = ref.nativeId, clientId = ref.clientId))
    }

    // ── 플랫폼 통합 (실제 구현) ───────────────────────────────────────────────

    AsyncFunction("openSettings") Coroutine { ->
      withContext(Dispatchers.Main) {
        WorkoutsIntents.openSettings(requireContext(), appContext.currentActivity)
      }
    }

    AsyncFunction("openStoreListing") Coroutine { ->
      withContext(Dispatchers.Main) {
        WorkoutsIntents.openStoreListing(requireContext(), appContext.currentActivity)
      }
    }
  }

  /** f113: 남의 루트를 읽기 위한 하드 전제조건. 백그라운드 읽기 권한은 도움이 되지 않는다. */
  private fun isForeground(): Boolean =
    gateway.processImportance() <= RealHealthConnectGateway.FOREGROUND_IMPORTANCE

  /**
   * `RouteAccess` 축약.
   *
   * 정본은 AOSP의 `getExerciseRouteReadAccessType`이지만 connect-client 1.1.0에는 그
   * 질의가 없다. 아래는 f110/f113/f114가 측정한 것에서 파생한 보수적 유도이며, 그 API에 닿을 수
   * 있게 되면 **그것으로 교체한다**. 이 값은 절대 캐시하지 않는다 — 앱은 자기가 쓴 루트에 대한
   * 접근도 잃을 수 있다(f114).
   */
  private fun routeAccess(granted: Set<String>, foreground: Boolean): String = when {
    granted.contains(WorkoutsPermissions.READ_EXERCISE_ROUTES) && foreground -> "all"
    granted.contains(WorkoutsPermissions.WRITE_EXERCISE_ROUTE) -> "own"
    else -> "perRoute"
  }
}

/** Android에는 sync-identifier 조회가 없다(설계 §3.2). 타입을 명시해야 reify가 된다. */
private val NO_EXISTING_WORKOUT: Map<String, Any?>? = null

/** `WorkoutWriteRecord`(브리지) -> `WorkoutWriteDto`(seam). expo 타입이 seam을 건너지 않게 한다. */
private fun WorkoutWriteRecord.toDto(): WorkoutWriteDto = WorkoutWriteDto(
  clientId = clientId,
  version = version.toLong(),
  activityTypeRaw = activityTypeRaw,
  startMs = startMs,
  endMs = endMs,
  utcOffsetMin = utcOffsetMin,
  timeZoneId = timeZoneId,
  pauses = pauses.map { it.toDto() },
  laps = laps.map { it.toDto() },
  distanceM = distanceM,
  activeEnergyKcal = activeEnergyKcal,
  elevationGainM = elevationGainM,
  steps = steps?.toLong(),
  heartRate = heartRate.map { it.toDto() },
  route = route.map { it.toDto() },
)
